kotlin-native/build-tools/src/main/kotlin/org/jetbrains/kotlin/ComponentsXmlResourceTransformerPatched.kt

```diff
@@ -0,0 +1,145 @@
+            if (requirements != null && requirements.childCount > 0) {
+                for (r in requirements.childCount - 1 downTo 0) {
+                    val requirement: Xpp3Dom = requirements.getChild(r)
+                    var requiredRole: String? = getValue(requirement, "role")
+                    requiredRole = getRelocatedClass(requiredRole, context)
+                    setValue(requirement, "role", requiredRole)
+                }
+            }
+            components[key] = component
+        }
+    }
+
+    override fun modifyOutputStream(os: ZipOutputStream, preserveFileTimestamps: Boolean) {
+        val data = transformedResource
+        val entry = ZipEntry(COMPONENTS_XML_PATH)
+        entry.time = TransformerContext.getEntryTimestamp(preserveFileTimestamps, entry.time)
+        os.putNextEntry(entry)
+        IOUtil.copy(data, os)
+        components.clear()
+    }
+
+    override fun hasTransformedResource(): Boolean {
+        return components.isNotEmpty()
+    }
+
+    private val transformedResource: ByteArray
+        get() {
+            val baos = ByteArrayOutputStream(1024 * 4)
+            val writer: Writer = WriterFactory.newXmlWriter(baos)
+            try {
+                val dom = Xpp3Dom("component-set")
+                val componentDom = Xpp3Dom("components")
+                dom.addChild(componentDom)
+                for (component in components.values) {
+                    componentDom.addChild(component)
+                }
+                Xpp3DomWriter.write(writer, dom)
+            } finally {
+                IOUtil.close(writer)
+            }
+            return baos.toByteArray()
+        }
+
+    companion object {
+        private const val COMPONENTS_XML_PATH = "META-INF/plexus/components.xml"
+        private fun getRelocatedClass(className: String?, context: TransformerContext): String? {
+            val relocators = context.relocators
+            val stats = context.stats
+            if (className != null && className.isNotEmpty() && relocators != null) {
+                for (relocator in relocators) {
+                    if (relocator.canRelocateClass(className)) {
+                        val relocateClassContext = RelocateClassContext(className, stats)
+                        return relocator.relocateClass(relocateClassContext)
+                    }
+                }
+            }
+            return className
+        }
+
+        private fun getValue(dom: Xpp3Dom, element: String): String {
+            val child: Xpp3Dom? = dom.getChild(element)
+            return if (child?.value != null) child.value else ""
+        }
+
+        private fun setValue(dom: Xpp3Dom, element: String, value: String?) {
+            val child: Xpp3Dom? = dom.getChild(element)
+            if (value == null || value.isEmpty()) {
+                return
+            }
+            child?.value = value
+        }
+    }
+}
```

Comment:
Could you, please, explain, why it is needed?
If it is not here by mistake, maybe it is a good idea to write about this file in the commit message and show this part of the commit to someone in the native team who understands the build system there.

Reply:
Oh, sorry. There was another transformer in this directory and I blindly placed it here. It is needed to correctly merge plexus components files. I think that `buildSrc` is maybe a better place for this file.

Reply:
Moved this file to buildSrc